geomesa-kafka/geomesa-kafka-datastore/src/main/scala/org/locationtech/geomesa/kafka/KafkaDataStoreSchemaManager.scala

```diff
@@ -0,0 +1,183 @@
+
+    // build the schema node
+    val typeName = featureType.getTypeName
+    val schemaPath: String = getSchemaPath(typeName)
+    if (zkClient.exists(schemaPath)) {
+      throw new IllegalArgumentException(s"Type $typeName already exists")
+    }
+
+    val data = SimpleFeatureTypes.encodeType(featureType)
+    createZkNode(schemaPath, data)
+
+    // build the topic node
+    createZkNode(getTopicPath(typeName), kfc.topic)
+
+    // build the replay config node (optional)
+    // build the topic node
+    kfc.replayConfig.foreach(r => createZkNode(getReplayConfigPath(typeName), ReplayConfig.encode(r)))
+
+    //create the Kafka topic
+    if(!AdminUtils.topicExists(zkClient, kfc.topic))
+      AdminUtils.createTopic(zkClient, kfc.topic, partitions, replication)
+
+    // put it in the cache
+    schemaCache.put(typeName, kfc)
+
+  }
+
+  def getFeatureConfig(typeName: String) : KafkaFeatureConfig = schemaCache.get(typeName)
+
+  /** Extracts the prepared-for-live [[SimpleFeatureType]] which the given prepared-for-replay
+    * ``replayType`` is based on.
+    */
+  def getLiveFeatureType(replayType: SimpleFeatureType): Option[SimpleFeatureType] = {
+
+    Try {
+      KafkaDataStoreHelper.extractLiveTypeName(replayType).map(schemaCache.get(_).sft)
+    }.getOrElse(None)
+  }
+
+  override def getNames: util.List[Name] = zkClient.getChildren(zkPath).asScala.map(
+    name => new NameImpl(name) : Name).asJava
+
+
+  override def removeSchema(typeName: Name): Unit = removeSchema(typeName.getLocalPart)
+
+  override def removeSchema(typeName: String): Unit = {
+
+    // clean up cache and zookeeper resources
+    schemaCache.invalidate(typeName)
+    zkClient.deleteRecursive(getSchemaPath(typeName))
+
+    // todo: what about other remote caches - do we set up a kafka listener looking for deletes?
+    // todo: what about kafka - who has the responsibility to clean the kafka topic up?
+
+  }
+
+  private def resolveTopicSchema(typeName: String): KafkaFeatureConfig = {
+
+    val schema = zkClient.readData[String](getSchemaPath(typeName)) //throws ZkNoNodeException if not found
+    val topic = zkClient.readData[String](getTopicPath(typeName)) // throws Zk...
+    val sft = SimpleFeatureTypes.createType(typeName, schema)
+    KafkaDataStoreHelper.insertTopic(sft, topic)
+    val replay = Option(zkClient.readData[String](getReplayConfigPath(typeName), true))
+    replay.map(KafkaDataStoreHelper.insertReplayConfig(sft, _))
+
+    KafkaFeatureConfig(sft)
+  }
+
+  def getSchemaPath(typeName: String): String = {
+    s"$zkPath/$typeName"
+  }
+  def getTopicPath(typeName: String): String = {
+    s"$zkPath/$typeName/Topic"
+  }
+  def getReplayConfigPath(typeName: String): String = {
+    s"$zkPath/$typeName/ReplayConfig"
+  }
+
+  private val zkClient = {
+    // zkStringSerializer is required - otherwise topics won't be created correctly
+    val ret = new ZkClient(zookeepers, Int.MaxValue, Int.MaxValue, ZKStringSerializer)
```

Comment:
we should really be closing this zkClient at some point... maybe override the 'destroy' method in datastore? (might be named something else, but i think there is a lifecycle method)